presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -476,7 +476,7 @@ private QueryInfo buildQueryInfo(SqlQueryScheduler scheduler)
 
         if (queryInfo.isFinalQueryInfo()) {
             // capture the final query state and drop reference to the scheduler
-            finalQueryInfo.compareAndSet(null, queryInfo);
+            stateMachine.setFinalQueryInfo(null, queryInfo);
```

Comment:
This code is basically:
```java
QueryInfo queryInfo = stateMachine.getQueryInfo(stageInfo); 
if (queryInfo.isFinalQueryInfo()) {
    // capture the final query state and drop reference to the scheduler
    stateMachine.setFinalQueryInfo(null, queryInfo);
    ....
```
Can we have the `getQueryInfo` method detect this case and internally set the `finalQueryInfo`?  If we can we will need to rename the method to something that doesn't imply it is a simple getter. Maybe `updateQueryInfo` or `buildQueryInfo`.